### ISSUE TITLE
Fix `mypy` in an environment where some dependencies are installed

### ISSUE THREAD

**Issue by crcrpar** [optuna/optuna#1801]
Currently, mypy fails if the environment like optuna/optuna:py3.7-dev has some integration dependencies installed as follows (though the command itself was ran the other day):

If this problem is solved, we can easily run mypy checks locally.

```shelll
~/.ghq/github.com/optuna/optuna yet-another-dumb-branch*
❯ docker run -it -v $(pwd):/workspaces optuna/optuna:py3.7-dev /bin/bash -c "black --check .; flake8 .; mypy ."

...

optuna/dashboard.py:235: error: Argument 1 to "coroutine" has incompatible type "Callable[[_DashboardApp], None]"; expected "Callable[..., Generator[Any, Any, <nothing>]]"
optuna/integration/chainer.py:13: error: Cannot assign to a type
optuna/integration/chainer.py:13: error: Incompatible types in assignment (expression has type "Type[object]", variable has type "Type[Extension]")
optuna/integration/chainer.py:74: error: Incompatible types in assignment (expression has type "Optional[Any]", variable has type "Union[float, Variable]")
optuna/integration/chainer.py:77: error: Argument 1 to "float" has incompatible type "Union[float, Variable]"; expected "Union[SupportsFloat, _SupportsIndex, str, bytes, bytearray]"
tests/test_logging.py:82: error: "CaptureFixture" has no attribute "at_level"
tests/test_logging.py:84: error: "CaptureFixture" has no attribute "text"
tests/test_logging.py:88: error: "CaptureFixture" has no attribute "at_level"
tests/test_logging.py:90: error: "CaptureFixture" has no attribute "text"
tests/test_logging.py:94: error: "CaptureFixture" has no attribute "at_level"
tests/test_logging.py:96: error: "CaptureFixture" has no attribute "text"
optuna/structs.py:78: error: Function is missing a type annotation for one or more arguments
tests/integration_tests/test_fastai.py:82: error: Cannot assign to a method
tests/integration_tests/test_chainer.py:47: error: Argument 3 to "ChainerPruningExtension" has incompatible type "TimeTrigger"; expected "Union[Tuple[int, str], IntervalTrigger, ManualScheduleTrigger]"
tests/integration_tests/test_chainer.py:89: error: Argument 1 to "__call__" of "ChainerPruningExtension" has incompatible type "_MockTrainer@83"; expected "Trainer"
tests/integration_tests/test_chainer.py:103: error: Argument 1 to "_observation_exists" of "ChainerPruningExtension" has incompatible type "_MockTrainer@97"; expected "Trainer"
tests/integration_tests/test_chainer.py:105: error: Argument 1 to "_observation_exists" of "ChainerPruningExtension" has incompatible type "_MockTrainer@97"; expected "Trainer"
tests/integration_tests/test_chainer.py:111: error: Argument 1 to "_observation_exists" of "ChainerPruningExtension" has incompatible type "_MockTrainer@97"; expected "Trainer"
tests/integration_tests/test_chainer.py:113: error: Argument 1 to "_observation_exists" of "ChainerPruningExtension" has incompatible type "_MockTrainer@97"; expected "Trainer"
tests/integration_tests/test_chainer.py:123: error: Argument 1 to "_get_float_value" of "ChainerPruningExtension" has incompatible type "List[<nothing>]"; expected "Union[float, Variable]"
Found 20 errors in 6 files (checked 190 source files)
```

**Comment by HideakiImamura** [comment]
Thanks for raising an important issue for developers. Now, I'm working on this issue relating to PR #1755. So, let me remove the `contribution-welcome` label.

**Comment by HideakiImamura** [comment]
## Note 
In PR #1804, to address the `error: Cannot assign to a type` issue for `optuna/integration/chainer.py:13:`, I just add `# type: ignore`.
```python
if not _imports.is_successful():
    Extension = object  # type: ignore # NOQA
```
However, as pointed by @toshihikoyanase in #1804, there are the same substitutions in other places such as
```python
# In optuna/integration/keras.py
if not _imports.is_successful():
    Callback = object  # NOQA

# In optuna/integration/allennlp.py
if _imports.is_successful():
    import _jsonnet
else:
    EpochCallback = object  # NOQA
```
I don't know how we avoid `error: Cannot assign to a type` without `# type: ignore`. This is future work for #1804. I will add the `contribution-welcome` label to this issue after merging #1804.

**Comment by toshihikoyanase** [comment]
@HideakiImamura Thank you for adding a comment about my review point. 

> I will add the contribution-welcome label to this issue after merging #1804.

If you make the problem `contribution-welcome`, how about creating an independent issue? It will clarify the focus and would be helpful for contributors.

**Comment by HideakiImamura** [comment]
@toshihikoyanase Thank you for the comment. I will create another issue after merging PR #1804.

**Comment by HideakiImamura** [comment]
Let me close this issue because #1804 has merged. I will create another issue to discuss the point as mentioned by @toshihikoyanase.
